llm/vicuna-llama-2/scripts/train.py

```diff
@@ -0,0 +1,337 @@
+        ret = dict(
+            input_ids=ret["input_ids"][0],
+            labels=ret["labels"][0],
+            attention_mask=ret["attention_mask"][0],
+        )
+        self.cached_data_dict[i] = ret
+
+        return ret
+
+
+def make_supervised_data_module(tokenizer: transformers.PreTrainedTokenizer,
+                                data_args) -> Dict:
+    """Make dataset and collator for supervised fine-tuning."""
+    dataset_cls = (LazySupervisedDataset
+                   if data_args.lazy_preprocess else SupervisedDataset)
+    rank0_print("Loading data...")
+
+    train_json = json.load(open(data_args.data_path, "r"))
+    train_dataset = dataset_cls(train_json, tokenizer=tokenizer)
+
+    if data_args.eval_data_path:
+        eval_json = json.load(open(data_args.eval_data_path, "r"))
+        eval_dataset = dataset_cls(eval_json, tokenizer=tokenizer)
+    else:
+        eval_dataset = None
+
+    return dict(train_dataset=train_dataset, eval_dataset=eval_dataset)
+
+
+class CheckpointCallback(transformers.TrainerCallback):
+
+    def on_save(self, args, state, control, **kwargs):
+        """Add complete indicator to avoid incomplete checkpoints."""
+        if state.is_world_process_zero:
+            ckpt_path = os.path.join(args.output_dir,
+                                     f'checkpoint-{state.global_step}')
+            with open(os.path.join(ckpt_path, 'complete'), 'w') as f:
+                f.write('')
+            print(f'Checkpoint {state.global_step} saved.')
+        torch.distributed.barrier()
+
+
+def cleanup_incomplete_checkpoints(output_dir):
+    """Remove incomplete checkpoints."""
+    checkpoints = list(pathlib.Path(output_dir).glob('checkpoint-*'))
+    checkpoints = [c for c in checkpoints if c.name.split('-')[-1].isdigit()]
+    checkpoints = sorted(checkpoints,
+                         key=lambda x: int(x.name.split('-')[-1]),
+                         reverse=True)
+    for checkpoint in checkpoints:
+        if not (checkpoint / 'complete').exists():
+            print(f'Removing incomplete checkpoint {checkpoint}')
+            shutil.rmtree(checkpoint)
+        else:
+            print(f'Using checkpoint {checkpoint}')
+            break
+
+
+def train():
+    global local_rank
+
+    parser = transformers.HfArgumentParser(
+        (ModelArguments, DataArguments, TrainingArguments))
+    model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+    local_rank = training_args.local_rank
+    if local_rank == 0:
+        cleanup_incomplete_checkpoints(training_args.output_dir)
+    torch.distributed.barrier()
+
+    # Check the existence of checkpoints in all processes
+    resume_from_checkpoint = False
+    checkpoints = list(
+        pathlib.Path(training_args.output_dir).glob('checkpoint-*'))
+    checkpoints = [c for c in checkpoints if c.name.split('-')[-1].isdigit()]
+    if checkpoints:
+        resume_from_checkpoint = True
```

Comment:
I was trying to solve the loss mismatch problem when recovering from the checkpoint.
IIUC, the program is running in an MPI style with the `torchrun`, the previous way for setting the `resume_from_checkpoint` only in the `local_rank==0`, will not propagate the variable to the other ranks.

Reply:
Nice! Maybe add "... All ranks must simultaneously resume from a checkpoint if it exists. Otherwise, upon recovery the model weights may not reload correctly, causing loss spikes."

Reply:
Sounds good! Will add it. Just tested the latest implementation, and it seems the loss spikes go away.